File: bench/langs/ledger.f90
! Keeps the running balance for one account.
module ledger
   implicit none
   integer, parameter :: LIMIT = 5000

contains

   function refuse(owner) result(left)
      character(*), intent(in) :: owner
      integer :: left
      call warn_owner(owner)
      left = 0
   end function refuse

   function commit_entry(owner, amount) result(left)
      character(*), intent(in) :: owner
      integer, intent(in) :: amount
      integer :: left
      call write_entry(owner, amount)
      left = amount
   end function commit_entry

   ! Bills the account and returns what is left.
   function charge(owner, amount) result(left)
      character(*), intent(in) :: owner
      integer, intent(in) :: amount
      integer :: left
      if (amount > LIMIT) then
         left = refuse(owner)
         return
      end if
      left = commit_entry(owner, amount)
   end function charge

end module ledger

File: bench/langs/ledger.zig
// Keeps the running balance for one account.

const LIMIT = 5000;

fn refuse(owner: []const u8, amount: i64) i64 {
    warnOwner(owner);
    return 0;
}

fn commitEntry(owner: []const u8, amount: i64) i64 {
    writeEntry(owner, amount);
    return amount;
}

// Bills the account and returns what is left.
pub fn charge(owner: []const u8, amount: i64) i64 {
    if (amount > LIMIT) {
        return refuse(owner, amount);
    }
    return commitEntry(owner, amount);
}
